DataDog/dd-trace-dotnet
# [Tracer] refactor trace serialization part 2: `_sampling_priority_v1`

## Summary of changes

Stop adding attribute `_sampling_priority_v1` as tags to in-memory `Span` instances. Instead, track its value in `TraceContext` and inject it into MessagePack during serialization.

## Reason for change

TL;DR performance and moving towards supporting trace agent endpoints 0.6/0.7

- consolidate trace-level metadata and remove duplication in spans
- offload more work to the serialization thread
- avoid adding the sampling priority tag to _all_ spans, resulting in smaller payloads
- make the `CommonTags` object (used in every auto-instrumentation span) smaller.
- Bonus: small code clean-up in `TraceChunkModel`

## Implementation details

- remove `CommonTags.SamplingPriority`
- inject `_sampling_priority_v1` tag during serialization
    - use the value from `TraceContext.SamplingPriority`, not from `CommonTags.SamplingPriority` (removed) or `SpanContext.SamplingPriority` (used only for propagation)
    - inject sampling priority into all spans that are eligible to be selected as "chunk root" by the trace agent
- cache the UTF-8 bytes for each sampling priority value for fast indexed lookup
- added `TraceContext.WriteClosedSpans()` to force partial flush from tests

## Test coverage

- update `TraceContextTests.cs` because it looked for the sampling priority tag
- add `SamplingPriorityTestsWithoutUpstreamService` and `SamplingPriorityTestsWithUpstreamService` to assert that the sampling priority tag is where we expect it when a trace is split into chunk in different ways
- TODO: add more tests (see Pierre's comment below)

## Other details
Follows from #3135.

Note for posterity: as of today (agent version 7.39.1), to find a trace chunk's sampling priority, the agent follows these steps, in order:
- find the span that the agent calls the "chunk root"
    - iterate backwards from the _end_ of the trace chunk, if any span has `parentID == 0`, it selects that span as the chunk root
    - find all spans that do not have a parent in the same trace chunk (I'm calling those "chunk orphans") and select one of them _non-deterministically_ as the chunk root
    - as a fallback, it selects the last span in the chunk. this should never happen in a valid trace chunk because it requires circular references to occur.
- if the selected chunk root has the sampling priority tag, it uses that one
- [agent versions >= 7.34.0] if the selected chunk root does _not_ have the sampling priority tag, iterate the trace chunk from the _beginning_ and use the first sampling priority it finds

Due to the order of steps above, tracers can help the trace agent find the sampling priority faster:
- place the span with `parentID == 0` at the end of the trace chunk (this already happens in .NET in most cases)
- if there is no span `parentID == 0`, add the sampling priority tag at least to all chunk orphans
- if the tracer cannot follow the steps above, add the sampling priority tag to the first span if the agent versions is >= 7.34.0, or to _all_ spans if the agent version is lower

Side note: `_dd.origin` follows similar logic, except it does _not_ have the fallback added in agent 7.34.0

See also
- `GetRoot()` at https://github.com/DataDog/datadog-agent/blob/7433618413122f1556cb681252ea01e426e7136c/pkg/trace/traceutil/trace.go#L53-L90
- `normalizeChunk()` at https://github.com/DataDog/datadog-agent/blob/7433618413122f1556cb681252ea01e426e7136c/pkg/trace/agent/normalizer.go#L148-L167
- the call to both at https://github.com/DataDog/datadog-agent/blob/7433618413122f1556cb681252ea01e426e7136c/pkg/trace/agent/agent.go#L240-L241
